warehouse/query-core/src/main/java/datawave/query/jexl/visitors/RewriteNullFunctionsVisitor.java

```diff
@@ -0,0 +1,485 @@
+import org.apache.commons.jexl2.parser.ASTSizeFunction;
+import org.apache.commons.jexl2.parser.ASTSizeMethod;
+import org.apache.commons.jexl2.parser.ASTStringLiteral;
+import org.apache.commons.jexl2.parser.ASTTernaryNode;
+import org.apache.commons.jexl2.parser.ASTTrueNode;
+import org.apache.commons.jexl2.parser.ASTUnaryMinusNode;
+import org.apache.commons.jexl2.parser.ASTVar;
+import org.apache.commons.jexl2.parser.ASTWhileStatement;
+import org.apache.commons.jexl2.parser.JexlNode;
+import org.apache.commons.jexl2.parser.JexlNodes;
+import org.apache.commons.jexl2.parser.ParserTreeConstants;
+import org.apache.commons.jexl2.parser.SimpleNode;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+import static datawave.query.jexl.functions.EvaluationPhaseFilterFunctions.EVAL_PHASE_FUNCTION_NAMESPACE;
+import static datawave.query.jexl.functions.EvaluationPhaseFilterFunctionsDescriptor.IS_NULL;
+
+/**
+ * This visitor prepares the query tree for the {@link IsNotNullPruningVisitor}. It should be run after QueryModel expansion.
+ * <p>
+ * It is similar to the {@link IsNotNullIntentVisitor} in that it rewrites certain filter functions into their ASTEq equivalents.
+ * <ul>
+ * <li><code>filter:isNull(FIELD)</code></li>
+ * <li><code>filter:isNull(FIELD_1 || FIELD_2)</code></li>
+ * <li><code>filter:isNotNull(FIELD)</code></li>
+ * <li><code>filter:isNotNull(FIELD_1 || FIELD_2)</code></li>
+ * </ul>
+ */
+public class RewriteNullFunctionsVisitor extends BaseVisitor {
+    
+    // used to rebuild flattened unions
+    private boolean rebuiltMultiFieldedFunction = false;
+    
+    private final String IS_NOT_NULL = "isNotNull";
+    
+    private RewriteNullFunctionsVisitor() {}
+    
+    /**
+     *
+     * @param node
+     *            an arbitrary JexlNode
+     * @return the original query tree
+     */
+    @SuppressWarnings("unchecked")
+    public static <T extends JexlNode> T rewriteNullFunctions(JexlNode node) {
+        node.jjtAccept(new RewriteNullFunctionsVisitor(), null);
+        return (T) node;
+    }
+    
+    /**
+     * This is required to produce a flattened query tree
+     *
+     * @param node
+     *            an ASTOrNode
+     * @param data
+     *            an object
+     * @return the same node
+     */
+    @Override
+    public Object visit(ASTOrNode node, Object data) {
+        node.childrenAccept(this, data);
+        
+        if (rebuiltMultiFieldedFunction) {
+            List<JexlNode> children = new ArrayList<>();
+            for (JexlNode child : JexlNodes.children(node)) {
+                JexlNode deref = JexlASTHelper.dereference(child);
+                if (deref instanceof ASTOrNode) {
+                    children.addAll(Arrays.asList(JexlNodes.children(deref)));
+                } else {
+                    children.add(child);
+                }
+            }
+            JexlNodes.children(node, children.toArray(new JexlNode[0]));
```

Comment:
Not something that needs to be done as part of this PR, but it would be nice to have a `JexlNodes.children(JexlNode, Collection)` method so that we don't have to create a new array in these cases.

Reply:
Created https://github.com/NationalSecurityAgency/datawave/issues/1517